src/eventarc/eventarc.ts

```diff
@@ -0,0 +1,287 @@
+  }
+
+  /**
+   * Publishes provided events to this channel. If channel was created with
+   * `allowedEventsTypes` and event type is not on that list, the event will
+   * be ignored.
+   * 
+   * The following CloudEvent fields will be auto-populated if not set:
+   *  * specversion - `1.0`
+   *  * id - uuidv4()
+   *  * source - will be populated with `process.env.EVENTARC_CLOUD_EVENT_SOURCE` and 
+   *             if not set an error will be thrown.
+   *  
+   * @param events - CloudEvent to publish to the channel.
+   */
+  public async publish(events: CloudEvent | CloudEvent[]): Promise<void> {
+    if (!Array.isArray(events)) {
+      events = [events as CloudEvent];
+    }
+    return this.publishToEventarcApi(
+      await this.resolvedChannelName,
+      events
+        .filter(e => typeof this.allowedEventTypes === 'undefined' || this.allowedEventTypes.includes(e.type))
+        .map(toCloudEventProtoFormat));
+  }
+
+  private async publishToEventarcApi(channel:string, events: CloudEvent[]): Promise<void> {
+    if (events.length === 0) {
+      return;
+    }
+    const request: HttpRequestConfig = {
+      method: 'POST',
+      url: `${EVENTARC_API}/${channel}:publishEvents`,
+      data: JSON.stringify({ events }),
+    };
+    return this.sendRequest<void>(request);
+  }
+
+  private sendRequest<T>(request: HttpRequestConfig): Promise<T> {
+    request.headers = FIREBASE_VERSION_HEADER;
+    return this.httpClient.send(request)
+      .then((resp) => {
+        return resp.data as T;
+      })
+      .catch((err) => {
+        throw this.toFirebaseError(err);
+      });
+  }
+
+  private toFirebaseError(err: HttpError): PrefixedFirebaseError {
+    if (err instanceof PrefixedFirebaseError) {
+      return err;
+    }
+
+    const response = err.response;
+    if (!response.isJson()) {
+      return new PrefixedFirebaseError(
+        'eventarc',
+        '' + response.status,
+        `Unexpected response with status: ${response.status} and body: ${response.text}`);
+    }
+
+    return new PrefixedFirebaseError(
+      'eventarc', 
+      'unknown-error',
+      `Unexpected response with status: ${response.status} and body: ${response.text}`);
+  }
+
+  private resolveChannelName(name: string): Promise<string> {
+    if (!name.includes('/')) {
+      const location = DEFAULT_CHANNEL_REGION;
+      const channelId = name;
+      return this.resolveChannelNameProjectId(location, channelId);
+    } else {
+      const match = CHANNEL_NAME_REGEX.exec(name);
+      if (match === null) {
+        throw new FirebaseEventarcError('invalid-argument', 'Invalid channel name format.');
+      }
+      const projectId = match[2];
+      const location = match[3];
+      const channelId = match[4];
+      if (projectId) {
+        return Promise.resolve('projects/' + projectId + '/locations/' + location + '/channels/' + channelId);
+      } else {
+        return this.resolveChannelNameProjectId(location, channelId);
```

Comment:
Let's make this a util function and decouple it from `utils.findProjectId(...)` that way we can cache the default project id for future instance `this.projectId = projectId` preventing repeated calls to `utils.findProjectId(...)`.

Reply:
Done.